runtime/alloc/alloc.c

```diff
@@ -3,177 +3,72 @@
-    if (next_superblock_ptr) {
-      *next_superblock_ptr = superblock_ptr;
-    }
-    memory_block_header *hdr = (memory_block_header *)superblock_ptr;
-    next_superblock_ptr = &hdr->next_superblock;
-    hdr->next_superblock = 0;
-  }
-  blocks_left--;
-  void* result = superblock_ptr;
-  superblock_ptr += BLOCK_SIZE;
-  return result;
-}
-
-static void freshBlock(struct semispace_info *space) {
-    char *nextBlock;
-    if (space->block_start == 0) {
-      nextBlock = megabyte_malloc();
-      space->first_block = nextBlock;
-      memory_block_header *nextHeader = (memory_block_header *)nextBlock;
-      nextHeader->next_block = 0;
-      nextHeader->semispace = space->semispace_id;
-    } else {
-      nextBlock = *(char**)space->block_start;
-      if (space->block != space->block_end) {
-        if (space->block_end - space->block == 8) {
-          *(uint64_t *)space->block = NOT_YOUNG_OBJECT_BIT; // 8 bit sentinel value
-        } else {
-          *(uint64_t *)space->block = space->block_end - space->block - 8; // 16-bit or more sentinel value
-        }
-      }
-      if (!nextBlock) {
-        MEM_LOG("Allocating new block for the first time in semispace %d\n", space->semispace_id);
-        nextBlock = megabyte_malloc();
-        *(char **)space->block_start = nextBlock;
-        memory_block_header *nextHeader = (memory_block_header *)nextBlock;
-        nextHeader->next_block = 0;
-        nextHeader->semispace = space->semispace_id;
-      }
-    }
-    space->block = nextBlock + sizeof(memory_block_header);
-    space->block_start = nextBlock;
-    space->block_end = nextBlock + BLOCK_SIZE;
-    MEM_LOG("New block at %p (remaining %zd)\n", space->block, BLOCK_SIZE - sizeof(memory_block_header));
-}
-
-static void* __attribute__ ((noinline)) doAllocSlow(size_t requested, struct semispace_info *space) {
-  MEM_LOG("Block at %p too small, %zd remaining but %zd needed\n", space->block, space->block_end-space->block, requested);
-  if (requested > BLOCK_SIZE - sizeof(memory_block_header)) {
-     return malloc(requested);
-  } else {
-    freshBlock(space);
-    void* result = space->block;
-    space->block += requested;
-    MEM_LOG("Allocation at %p (size %zd), next alloc at %p (if it fits)\n", result, requested, space->block);
-    return result;
-  }
-}
-
-static inline __attribute__ ((always_inline)) void* doAlloc(size_t requested, struct semispace_info *space) {
-  if (space->block + requested > space->block_end) {
-    return doAllocSlow(requested, space);
-  }
-  void* result = space->block;
-  space->block += requested;
-  MEM_LOG("Allocation at %p (size %zd), next alloc at %p (if it fits)\n", result, requested, space->block);
-  return result;
+  freeAllMemory();
+  arenaReset(&youngspace);
+  arenaReset(&oldspace);
 }
 
 __attribute__ ((always_inline)) void* koreAlloc(size_t requested) {
-  return doAlloc(requested, &fromspace);
+  return arenaAlloc(&youngspace, requested);
```

Comment:
similarly here, is arenaAlloc getting inlined?

Reply:
Same here. In fact, `doAlloc` is also getting inlined. There only calls to `doAllocSlow` in the binary.